keps/sig-api-machinery/3937-declarative-validation/README.md

```diff
@@ -0,0 +1,1189 @@
+    the same errors for validation_test.go.
+  - Enable e2e testing testing of the API group so that each e2e test run includes
+    the API group in `DeclarativeValidationEnabled` and checks for any differences
+    reported in logs between declarative validation and legacy validation and
+    fails the e2e test run if any are found.
+  - API reviewers approve the change.
+- ...Time passes. All API validation changes are deamed stable...
+- In a single PR:
+  - The `DeclarativeValidationEnabled` is defaulted to true.
+  - The `LegacyValidationEnabled` is defaulted to false.
+- ...Deprecation of legacy validation is announced...
+- ...Deprecation wait period passes...
+- The legacy API is deleted and can no longer be enabled/disabled via the
+  feature gates.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- Risk: Declarative validation adds significant latency to API request handling.
+  - Potential causes:
+    - Validation versioned types introduces extra conversions.
+      - Why this might be OK: Requests are received as the versioned type, so it
+        should be feasible to avoid extra conversions.
+    - CEL expressions are significantly more expensive to run.
+      - Why this might be OK:
+        - CEL expressions are expected to account for 30% of all
+          validation rules, with the rest expressed as simple value validations
+          (required checks, enum value checks) that do not require CEL
+          evaluation.
+        - CEL evaluation is relatively efficient, we will gather detailed
+          benchmarks baselined against native types to demonstrate this.
+        - If CEL becomes a bottleneck, we have potential mitigations:
+          - Generate Go code from CEL cel expressions.
+          - Hand write validation code for select validations and call those
+            directly (or from CEL).
+    - A general purpose validator must (a) convert to unstructured or (b) use
+      reflection.
+      - Why this might be OK: SMD's
+        [value](https://github.com/kubernetes-sigs/structured-merge-diff/tree/master/value)
+        wrappers have already been benchmarked and shown to have low overhead.
+- Risk: Migration introduces breaking change to API validation.
+  - Mitigations: See above migration plan, which includes numerous cross checks
+    to prevent mistakes from slipping through.
```

Comment:
Today validation for all the versions are done on internal, we might have discrepencies between versions now (since we'll require N validation instead of 1). We have to be sure that all the validations do the exact same thing.